events/ri/src/main/java/org/projectnessie/events/ri/kafka/KafkaEventSubscriber.java

```diff
@@ -0,0 +1,379 @@
+            referenceName(upstreamEvent.getReference()),
+            null,
+            upstreamEvent.getHashAfter());
+    fireEvent(upstreamEvent, downstreamEvent);
+  }
+
+  @Override
+  public void onReferenceUpdated(org.projectnessie.events.api.ReferenceUpdatedEvent upstreamEvent) {
+    ReferenceEvent downstreamEvent =
+        new ReferenceEvent(
+            ReferenceEventType.REASSIGNED,
+            upstreamEvent.getId(),
+            referenceName(upstreamEvent.getReference()),
+            upstreamEvent.getHashBefore(),
+            upstreamEvent.getHashAfter());
+    fireEvent(upstreamEvent, downstreamEvent);
+  }
+
+  @Override
+  public void onReferenceDeleted(org.projectnessie.events.api.ReferenceDeletedEvent upstreamEvent) {
+    ReferenceEvent downstreamEvent =
+        new ReferenceEvent(
+            ReferenceEventType.DELETED,
+            upstreamEvent.getId(),
+            referenceName(upstreamEvent.getReference()),
+            upstreamEvent.getHashBefore(),
+            null);
+    fireEvent(upstreamEvent, downstreamEvent);
+  }
+
+  @SuppressWarnings("FutureReturnValueIgnored")
+  private void fireEvent(
+      org.projectnessie.events.api.Event upstreamEvent, SpecificRecord downstreamEvent) {
+    String key = eventKey(upstreamEvent, downstreamEvent);
+    ProducerRecord<String, Object> record = new ProducerRecord<>(topicName, key, downstreamEvent);
+    addCommonHeaders(upstreamEvent, record);
+    // fire and forget – we don't wait for the result
+    producer.send(record);
+  }
+
+  private void addCommonHeaders(
+      org.projectnessie.events.api.Event upstreamEvent, ProducerRecord<?, ?> record) {
+    Headers headers = record.headers();
+    Header.SPEC_VERSION.addValue(headers, serverSpecVersion);
+    Header.EVENT_TYPE.addValue(headers, upstreamEvent.getType().name().getBytes(UTF_8));
+    Header.REPOSITORY_ID.addValue(headers, upstreamEvent.getRepositoryId().getBytes(UTF_8));
+    upstreamEvent
+        .getEventInitiator()
+        .ifPresent(user -> Header.INITIATOR.addValue(headers, user.getBytes(UTF_8)));
+    Header.EVENT_CREATION_TIME.addValue(
+        headers, upstreamEvent.getEventCreationTimestamp().toString().getBytes(UTF_8));
+  }
+
+  /**
+   * Create a key for the downstream event. This is a combination of the repository ID and the
+   * reference name. This allows the topic to be partitioned by repository ID and reference name,
+   * which should help with downstream consumers that want to process events for a given reference
+   * in order.
+   *
+   * <p>Note: strict ordering of events is not guaranteed. If necessary, consider using Kafka
+   * Streams.
+   */
+  private static String eventKey(Event upstreamEvent, SpecificRecord downstreamEvent) {
+    String repositoryId = upstreamEvent.getRepositoryId();
+    String referenceName;
+    if (downstreamEvent instanceof CommitEvent) {
+      referenceName = ((CommitEvent) downstreamEvent).getTargetReference();
```

Comment:
Why the naming inconsistency? `getTargetReference` vs `getReference` below.

Reply:
In Nessie's Events API, a `CommitEvent` inherits from `CommittingEvent` where 2 methods are defined: `getSourceReference()` and `getTargetReference()`. In the case of a commit, both references are always the same (but they are different for merges and transplants). 

Whereas a `ReferenceEvent` has a distinct `getReference()` method, because it's always about one single reference.